### ISSUE TITLE
run_maintenance() fails after upgrading to 2.2.0

### ISSUE THREAD

**Issue by jfrost** [keithf4/pg_partman#71]
After upgrading to 2.2.0, the run_maintenance() function fails with:

```
# select partman.run_maintenance();

ERROR:  function show_partition_name(text, text) does not exist
CONTEXT: PL/pgSQL function partman.run_maintenance(text,boolean,boolean,boolean) line 167 at SQL statement
DETAIL:
HINT: No function matches the given name and argument types. You might need to add explicit type casts.
```

**Comment by keithf4** [comment]
Not sure how that could've happened. What's the result of running `\dx+ pg_partman` in psql?
```
keith=# \dx+ pg_partman 
                                        Objects in extension "pg_partman"
                                               Object Description                                                
-----------------------------------------------------------------------------------------------------------------
 function partman.apply_constraints(text,text,boolean,bigint,boolean)
 function partman.apply_foreign_keys(text,text,bigint,boolean)
 function partman.check_name_length(text,text,boolean)
 function partman.check_parent()
 function partman.check_partition_type(text)
 function partman.check_subpartition_limits(text,text)
 function partman.check_subpart_sameconfig(text)
 function partman.check_version(text)
 function partman.create_function_id(text,bigint)
 function partman.create_function_time(text,bigint)
 function partman.create_parent(text,text,text,text,text[],integer,boolean,text,boolean,boolean,boolean,boolean)
 function partman.create_partition_id(text,bigint[],boolean)
 function partman.create_partition_time(text,timestamp without time zone[],boolean)
 function partman.create_sub_parent(text,text,text,text,text[],integer,text,boolean,boolean,boolean,boolean)
 function partman.create_trigger(text)
 function partman.drop_constraints(text,text,boolean)
 function partman.drop_partition_id(text,bigint,boolean,boolean,text)
 function partman.drop_partition_time(text,interval,boolean,boolean,text)
 function partman.partition_data_id(text,integer,integer,numeric,text)
 function partman.partition_data_time(text,integer,interval,numeric,text)
 function partman.reapply_privileges(text)
 function partman.run_maintenance(text,boolean,boolean,boolean)
 function partman.show_partition_name(text,text)
 function partman.show_partitions(text,text)
 function partman.undo_partition_id(text,integer,bigint,boolean,numeric)
 function partman.undo_partition(text,integer,boolean,boolean,numeric)
 function partman.undo_partition_time(text,integer,interval,boolean,numeric)
 table partman.custom_time_partitions
 table partman.part_config
 table partman.part_config_sub
 type partman.check_parent_table
(31 rows)
```

**Comment by jfrost** [comment]
```
                                        Objects in extension "pg_partman"
                                               Object Description
-----------------------------------------------------------------------------------------------------------------
 function partman.apply_constraints(text,text,boolean,bigint,boolean)
 function partman.apply_foreign_keys(text,text,bigint,boolean)
 function partman.check_name_length(text,text,boolean)
 function partman.check_parent()
 function partman.check_partition_type(text)
 function partman.check_subpart_sameconfig(text)
 function partman.check_subpartition_limits(text,text)
 function partman.check_version(text)
 function partman.create_function_id(text,bigint)
 function partman.create_function_time(text,bigint)
 function partman.create_parent(text,text,text,text,text[],integer,boolean,text,boolean,boolean,boolean,boolean)
 function partman.create_partition_id(text,bigint[],boolean)
 function partman.create_partition_time(text,timestamp without time zone[],boolean)
 function partman.create_sub_parent(text,text,text,text,text[],integer,text,boolean,boolean,boolean,boolean)
 function partman.create_trigger(text)
 function partman.drop_constraints(text,text,boolean)
 function partman.drop_partition_id(text,bigint,boolean,boolean,text)
 function partman.drop_partition_time(text,interval,boolean,boolean,text)
 function partman.partition_data_id(text,integer,integer,numeric,text)
 function partman.partition_data_time(text,integer,interval,numeric,text)
 function partman.reapply_privileges(text)
 function partman.run_maintenance(text,boolean,boolean,boolean)
 function partman.show_partition_name(text,text)
 function partman.show_partitions(text,text)
 function partman.undo_partition(text,integer,boolean,boolean,numeric)
 function partman.undo_partition_id(text,integer,bigint,boolean,numeric)
 function partman.undo_partition_time(text,integer,interval,boolean,numeric)
 table partman.custom_time_partitions
 table partman.part_config
 table partman.part_config_sub
 type partman.check_parent_table
(31 rows)
```

**Comment by jfrost** [comment]
BTW, this is on 9.4.4 if that makes any difference.

**Comment by keithf4** [comment]
I'm really not sure how that error could be getting thrown. The function it says is missing is right there in your list. Does `\df partman.` show that it's there too? What version of pg_partman did you upgrade from?

**Comment by jfrost** [comment]
Previous version was 2.1.0.

Looks like it's a search_path problem. If I do this:

`set search_path = "$user",public,partman;`

then it works fine.

In 2.1.0, I didn't have to do that though.

**Comment by keithf4** [comment]
Ahh I see the issue. I forgot to use the @extschema@ macro in the run_maintenance() calls to that function. I'll try and get a patch out ASAP.

**Comment by jfrost** [comment]
Great!

Thanks for the quick response!

**Comment by keithf4** [comment]
Also realized why I didn't hit this bug in testing. You must not have pg_jobmon installed. If it is, it resets the search path to include pg_partman's schema along with pg_jobmon's.

**Comment by jfrost** [comment]
Totally correct!

Just using a cron job to call it.




-- 

*Jeff Frost*
Senior Director - DevOps  |  *Procore*
6309 Carpinteria Ave  |  Carpinteria, CA 93013
+1 805-456-5176  |  jeff.frost@procore.com  |  LinkedIn

*Watch our 1 minute overview <https://procore.wistia.com/medias/s7orcyx2q7>*

[image: Procore]
www.procore.com  |  LinkedIn
<http://www.linkedin.com/company/procore-technologies>  |  Facebook
<http://www.facebook.com/procore.tech>  |  Twitter
<http://www.twitter.com/procoretech>  |  YouTube
<http://www.youtube.com/user/ProcoreVideo>

Procore Technologies, Inc., the world's number one most widely used
construction management software, helps firms drastically increase project
efficiency and accountability by streamlining and mobilizing project
communication and documentation. Users manage all types of construction
projects including industrial plants, office buildings, apartment
complexes, university facilities, retail centers, and more.

**Comment by keithf4** [comment]
Fixed this issue in 2.2.1. Let me know if it's working ok for you know.

Just some FYI:
- pg_jobmon isn't for running maintenance or anything. If you're running pg_partman in production, I'd recommend looking into putting it in place. It can alert you to maintenance failing to run on schedule and provide an audit trail as to where and why it failed. It's also useful if you have other critical functions that must run in your database and you want to do the same for them.
- Since you're on 9.4, you don't actually need cron anymore for normal maintenance. If you're calling it on individual partition sets, then you still need cron, but otherwise you can use the background worker for that now.

**Comment by jfrost** [comment]
Sorry it took me a bit to get that server updated. It works perfectly now.
